Re-wire the `Users` entity to use RTK Query under the hood

nemanjaglumac opened the pull request:
### What does this PR accomplish?
- It adds new RTK Query api endpoints
- It migrates `Users` entity to use new RTK Query endpoints under the hood
- Even though this was not a part of this task/milestone, it also migrates `updatePassword` from the hook to the new RTK mutation
- It removes some unused `userApi` keys

Resolves https://github.com/metabase/metabase/issues/40812.

### What was out of the scope?
- Creating new RTK Query endpoints for the remaining `userAPi` entries:
    - list
    - current
    - update_qbnewb

They are not used by entities, and will be migrated in the milestone 2 of #40788

nemanjaglumac committed:
Define the types for the `create` user request

nemanjaglumac committed:
First draft for the create user API using RTK Query

nemanjaglumac committed:
Export RTK user API

nemanjaglumac committed:
Define explicit `create` api for the user entity

nemanjaglumac committed:
Define `updatePassword` mutation

nemanjaglumac committed:
Wire users entity `resetPasswordManual` to use RTK under the hood

nemanjaglumac committed:
Use `updatePassword` mutation

nemanjaglumac committed:
Wire users list calls

nemanjaglumac committed:
Remove unused `update_password` from services

nemanjaglumac committed:
Fix the dispatch used to list users

nemanjaglumac committed:
Dispatch the action type and the payload

nemanjaglumac committed:
Wire `deactivateUser` to work with RTK

nemanjaglumac committed:
Wire `reactivateUser` to work with RTK

nemanjaglumac committed:
Delete unused entries from `UserApi`

nemanjaglumac committed:
Convert the reducer to use `switch` statement

nemanjaglumac committed:
Remove unused user `send_invite` endpoint from FE

nemanjaglumac committed:
Wire `resetPasswordEmail` in users entity to work with RTK

nemanjaglumac commented on frontend/src/metabase/entities/users.js:
```diff
@@ -1,28 +1,28 @@
-const getRecipientsList = GET("/api/user/recipients");
+const getUserList = (query = {}, dispatch) =>
+  entityCompatibleQuery(query, dispatch, userApi.endpoints.listUsers);
+const getRecipientsList = dispatch =>
+  entityCompatibleQuery({}, dispatch, userApi.endpoints.listUserRecipients);
```
Maybe it would've been cleaner to pass both the `query` and the `dispatch` to both functions?

nemanjaglumac replied:
https://github.com/metabase/metabase/pull/41312/commits/d09565283d505039356bc19fbcf1cc10cb49b933

kamilmielnik commented on frontend/src/metabase/entities/users.js:
```diff
@@ -80,50 +85,74 @@ const Users = createEntity({
+          "People Admin",
+          "Trigger User Password Reset",
+        );
+        await await entityCompatibleQuery(
```
```suggestion
        await entityCompatibleQuery(
```

kamilmielnik commented on frontend/src/metabase/entities/users.js:
```diff
@@ -35,8 +35,14 @@ const Users = createEntity({
   api: {
-    list: ({ recipients = false, ...args }) =>
-      recipients ? getRecipientsList() : getUserList(args),
+    list: ({ recipients = false, ...args }, dispatch) =>
```
We should override all 5 default endpoints (i.e. also `get`, `update`, `delete`).
For the ones that are unused we should `throw` to prevent adding new code using legacy stuff.

nemanjaglumac replied:
Done.
Thanks

kamilmielnik replied:
@nemanjaglumac please don't forget about `delete` :pray: 

It should use `deactivateUser` endpoint.

kamilmielnik commented on frontend/src/metabase/api/user.ts:
```diff
@@ -0,0 +1,68 @@
+
+export const userApi = Api.injectEndpoints({
+  endpoints: builder => ({
+    listUsers: builder.query<ListUsersResponse, ListUsersRequest | null>({
```
```suggestion
    listUsers: builder.query<ListUsersResponse, ListUsersRequest | void>({
```

kamilmielnik commented on frontend/src/metabase/api/user.ts:
```diff
@@ -0,0 +1,68 @@
+      }),
+      // FIXME: providesTags is needed
+    }),
+    listUserRecipients: builder.query({
```
Please type request & response

nemanjaglumac replied:
https://github.com/metabase/metabase/pull/41312/commits/baa9d1bf04a8ce5ee6c8fc44818d021819eaf31f

kamilmielnik commented on frontend/src/metabase/api/user.ts:
```diff
@@ -0,0 +1,68 @@
+      }),
+      // FIXME: invalidatesTags is maybe needed?
+    }),
+    deactivateUser: builder.query<void, { id: UserId }>({
```
Other new endpoints would just accept `UserId`, not `{id: UserId}`.
Let's make this consistent?
Same goes for `reactivateUser`.

nemanjaglumac replied:
https://github.com/metabase/metabase/pull/41312/commits/e9498124a1bb676d2d563f51d59168796e597a56 and then https://github.com/metabase/metabase/pull/41312/commits/6369c71bf71923a22765ed14f54f3ae4592d6ecc

kamilmielnik commented on frontend/src/metabase/entities/users.js:
```diff
@@ -80,50 +85,74 @@ const Users = createEntity({
+        dispatch({ type: PASSWORD_RESET_EMAIL });
+      },
+    resetPasswordManual:
+      async ({ id }, password = generatePassword()) =>
```
Is this function `async`?

nemanjaglumac replied:
It was/is
https://github.com/metabase/metabase/pull/41312/commits/4bf0e0e504881a2b9fc29eda6761359bb138cf99#diff-4f46204c4fd684bc8b51d8c6f9a2d791ce065787b6de8dce6808772d1d02f5c8L103

kamilmielnik reviewed (commented):
Looks good.
Some tests are failing.
You can use other entity endpoints (e.g. cards) as a reference implementation for tags.

romeovs commented on frontend/src/metabase/api/user.ts:
```diff
@@ -0,0 +1,68 @@
+} from "metabase-types/api";
+
+import { Api } from "./api";
+// import { listTag, invalidateTags } from "./tags";
```
Remove the commented out code.

```suggestion
```

kamilmielnik replied:
We'll need these for tags which are not implemented yet (see "FIXME" comments below)

nemanjaglumac commented on frontend/src/metabase/account/password/actions.ts:
```diff
@@ -18,11 +15,3 @@ export const validatePassword = async (password: string) => {
   }
 };
-
-export const updatePassword = async (user: User, data: UserPasswordData) => {
```
I am aware, but I thought it might be possible to clean things up as I go (until I realized there'll be much more of this).

nemanjaglumac replied:
Oh boy, did it come back to bite me!
I spent several hours trying to debug why is the form ignoring the error! The solution is here:
https://github.com/metabase/metabase/pull/41312/commits/e6f6487d7e74c81d1d9cb817643a5d8db01f68b0

nemanjaglumac commented on frontend/src/metabase-types/api/user.ts:
```diff
@@ -79,3 +79,54 @@ export type UserLoginHistoryItem = {
+  | "login_attributes"
+>;
+
+export type CreateUserResponse =
```
I don't think it's mapping 1:1. I'll double check and will tidy up as I go.

nemanjaglumac replied:
It is missing `updated_at` but I switched back to `User` as you suggested.
b5a0bc62007007168a24eeabec31f03950748a75

Will investigate why other types don't have this key.

nemanjaglumac committed:
Provide cache invalidation for the `userApi`

nemanjaglumac committed:
Add `provideUserListTags` helper

nemanjaglumac committed:
Pass `email` as a sole argument

nemanjaglumac commented on frontend/src/metabase/api/session.ts:
```diff
@@ -14,7 +14,15 @@ export const sessionApi = Api.injectEndpoints({
         body: { token },
       }),
     }),
+    forgotPassword: builder.query<void, { email: string }>({
```
https://github.com/metabase/metabase/pull/41312/commits/70a66de2c8b04b3e4cc61b41f1e6daa9bf05cf11

nemanjaglumac committed:
Throw on `get` a user

This endpoint hasn't been used in a long itme.
Removed in https://github.com/metabase/metabase/pull/41334/files#diff-241e0afa11f03a0942eb322aa47695e58496379b0b6d42e6e59ecf6a20fe623cL448

nemanjaglumac committed:
Use RTK Query to `updateUser`

nemanjaglumac committed:
Fix wrong syntax

nemanjaglumac committed:
Use RTK Query to `getUser`

It was used by `.load` and `fetch` after all.

nemanjaglumac committed:
Export `useUpdateUserMutation`

nemanjaglumac committed:
Merge branch 'master' into rtk-query-user-entity

nemanjaglumac committed:
Refactor `getRecipientsList`

nemanjaglumac committed:
Fix response type for the `getUser`

nemanjaglumac committed:
Pass user id as a sole argument instead of in an object

nemanjaglumac committed:
Fix users entity `get` method

nemanjaglumac committed:
Fix the user update password form

nemanjaglumac committed:
Fix unit tests

nemanjaglumac committed:
Fix deactivate and reactivate query invocations

See e949812.

nemanjaglumac committed:
Adjust types according to the feedback

nemanjaglumac commented on frontend/src/metabase-types/api/user.ts:
```diff
@@ -79,3 +79,54 @@ export type UserLoginHistoryItem = {
 
 export type UserLoginHistory = UserLoginHistoryItem[];
+
+export type CreateUserRequest = Pick<
```
b5a0bc62007007168a24eeabec31f03950748a75

kamilmielnik commented on frontend/src/metabase/api/tags/utils.ts:
```diff
@@ -326,3 +326,9 @@ export function provideTimelineTags(
   return [idTag("user", user.id)];
 }
+
+export function provideUserListTags(
```
These functions should be ordered alphabetically, i.e. this one should be defined before `provideUserTags`.

kamilmielnik commented on frontend/src/metabase/api/user.ts:
```diff
@@ -0,0 +1,100 @@
+        body,
+      }),
+      providesTags: response =>
+        response?.data ? provideUserListTags(response.data) : [],
```
```suggestion
        response ? provideUserListTags(response.data) : [],
```

nemanjaglumac replied:
Nope!
There were 10 different FE unit tests that were failing.

`response` may or may not contain `data`.

kamilmielnik replied:
> `response` may or may not contain `data`.

In that case either `ListUsersResponse` typing is incorrect or mocks are incorrectly set up in unit tests.
We should fix that.

kamilmielnik commented on frontend/src/metabase/api/user.ts:
```diff
@@ -0,0 +1,100 @@
+        url: "/api/user/recipients",
+      }),
+      providesTags: response =>
+        response?.data ? provideUserListTags(response.data) : [],
```
```suggestion
        response ? provideUserListTags(response.data) : [],
```

kamilmielnik reviewed (dismissed):
Good job! :+1: 

`delete` endpoint is not mapped to `deactivateUser` in entity framework - please update it before merging.

nemanjaglumac committed:
Order utils alphabetically

nemanjaglumac committed:
Throw on `delete`

nemanjaglumac committed:
Revert "Fix unit tests"

This reverts commit 2c329e836427010ad08bfd0511eb1a9df4ba95d6.

nemanjaglumac committed:
Fix FE unit test mocks

nemanjaglumac committed:
Rename imports to use plural for consistency

nemanjaglumac committed:
Fix one last remaining FE unit test
